chain/stateGenerator.go

```diff
@@ -62,16 +66,62 @@ func spendTransaction(states *db.StateDB, tx *Transaction, totalFee common.Fixed
 		nonceSender := accSender.GetNonce()
 		accSender.SetNonce(nonceSender + 1)
 		states.SetAccount(pg, accSender)
+	case GenerateIDType:
+		genID := pl.(*GenerateID)
+		pg, _ := common.ToCodeHash(tx.Programs[0].Code)
+		accSender := states.GetOrNewAccount(pg)
+		amountSender := accSender.GetBalance()
+		accSender.SetBalance(amountSender - common.Fixed64(genID.RegistrationFee) - common.Fixed64(tx.UnsignedTx.Fee))
+		nonceSender := accSender.GetNonce()
+		accSender.SetNonce(nonceSender + 1)
+		accSender.SetID(crypto.Sha256ZeroHash)
+		states.SetAccount(pg, accSender)
+
+		donationAddress, err := common.ToScriptHash(config.DonationAddress)
+		if err != nil {
+			return err
+		}
+		donationAccount := states.GetOrNewAccount(donationAddress)
+		amount := donationAccount.GetBalance()
+		donationAccount.SetBalance(amount + common.Fixed64(genID.RegistrationFee))
+		states.SetAccount(donationAddress, donationAccount)
+
 	}
 
 	return nil
 }
 
-func GenerateStateRoot(txs []*Transaction) common.Uint256 {
+func GenerateStateRoot(txs []*Transaction, height uint32, consensusData uint64) common.Uint256 {
```

Comment:
GenerateStateRoot() and spendTransaction() share lots of code with persist(). What about put them in the same function?

Reply:
I will optimize it later.